src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using Godot;
+
+public class ThriveInputEventWithModifiers : ICloneable
+{
+    public ThriveInputEventWithModifiers()
+    {
+    }
+
+    public ThriveInputEventWithModifiers(InputEventWithModifiers @event)
+    {
+        Control = @event.Control;
+        Alt = @event.Alt;
+        Shift = @event.Shift;
+        switch (@event)
+        {
+            case InputEventKey inputKey:
+                Type = InputType.Key;
+                Code = inputKey.Scancode;
+                break;
+            case InputEventMouseButton inputMouse:
+                Type = InputType.MouseButton;
+                Code = (uint)inputMouse.ButtonIndex;
+                break;
+        }
+    }
+
+    public enum InputType
+    {
+        Key,
+        MouseButton,
+    }
+
+    public bool Control { get; set; }
+    public bool Alt { get; set; }
+    public bool Shift { get; set; }
+    public InputType Type { get; set; }
+    public uint Code { get; set; }
+
+    /// <summary>
+    ///   Creates a string for the button to show.
+    /// </summary>
+    /// <returns>
+    ///   A human readable string.
+    /// </returns>
+    public override string ToString()
+    {
+        var text = string.Empty;
+
+        if (Control)
+            text += "Control+";
+        if (Alt)
+            text += "Alt+";
+        if (Shift)
+            text += "Shift+";
+        if (Type == InputType.Key)
+        {
+            var enumEntry = (KeyList)Code;
+            text += enumEntry.ToString();
+        }
+        else if (Type == InputType.MouseButton)
+        {
+            text += Code switch
+            {
+                1 => "Left mouse",
+                2 => "Right mouse",
+                3 => "Middle mouse",
+                4 => "Wheel up",
+                5 => "Wheel down",
+                6 => "Wheel left",
+                7 => "Wheel right",
+                8 => "Special 1 mouse",
+                9 => "Special 2 mouse",
+                _ => "Unknown mouse",
+            };
+        }
+
+        return text;
+    }
+
+    public InputEventWithModifiers ToGodotObject()
+    {
+        InputEventWithModifiers res = Type switch
```

Comment:
I think `var` would work here. Also spell out `result`

Reply:
var does not work here, because I have to specify the common type of the switch results

Reply:
Ah, okay. I'm still not experienced enough with C# it seems, almost every time I review your PRs I come across some new syntax.